### PR TITLE
Update iqtree

### PR DESCRIPTION
**Pull request by andrewsanchez**
We encountered an inscrutable `UnsatisfiableError` from conda after trying to resolve a complicated environment on osx.  [See the integration-darwin step here](https://busywork.qiime2.org/teams/main/pipelines/master-distribution/jobs/integration/builds/163).

After noticing this unexpected behavior in our CI system, we started examining our conda environments to try and understand the source of the issue.  

To summarize our findings (thanks, @ebolyen and @thermokarst 🕵! ):
- scikit-learn had been updated to a new version
- scikit-learn started being pulled from defaults, instead of conda-forge
- the defaults channel version of sklearn leaves `llvm-openmp` unconstrained. Our release build chose version 8 from conda-forge, even though version 9 had been available for months
- the conda-forge version of sklearn sets that same library to >=9, which apparently we couldn't find a resolution for, or the unconstrained openmp would have been 9 in the release (as it was available)
- it appeared that `llvm-openmp` is versioned in lockstep with `clangdev`, `llvmdev` , and `openmp`
- 9.0.1 drops the dependency on libcxx

This led us to suspect that packages depending on `llvm-openmp` could be the culprit.

We discovered that, by removing a specific package which depended on `iqtree`, conda could resolve the environment.  `iqtree` was a culprit due using `llvm-openmp@8.0.1` and `openmp@8.0.1`.

This PR is an attempt to avoid unsolvable environments in packages depending on `iqtree`.  Unfortunately, we had very little success running tests locally and are pushing up the first iteration of changes in order to have a look at the circleci results.  Notably, `circleci build` failed *after only incrementing the recipe build number,* without any other changes to `meta.yaml`.

 We had a look at the [spades recipe](https://github.com/bioconda/bioconda-recipes/blob/05e6af43cea7ed46a4f05d63d1d16fc49f7643ac/recipes/spades/meta.yaml) for guidance on configuring this recipe.

----

> Bioconda requires reviews prior to merging pull-requests (PRs). To facilitate this, once your PR is passing tests and ready to be merged, please add the `please review & merge` label so other members of the bioconda community can have a look at your PR and either make suggestions or merge it.

* [x] I have read the [guidelines for bioconda recipes](https://bioconda.github.io/contributor/guidelines.html).
* [ ] This PR adds a new recipe.
* [x] AFAIK, this recipe **is directly relevant to the biological sciences**
      (otherwise, please submit to the more general purpose [conda-forge channel](https://conda-forge.org/docs/)).
* [x] This PR updates an existing recipe.
* [ ] This PR does something else (explain below).

> Everyone has access to the following BiocondaBot commands, which can be given in a comment:
>
>  * `@BiocondaBot please update` will cause the BiocondaBot to merge the master branch into a PR
>  * `@BiocondaBot please add label` will add the `please review & merge` label.
>  * `@BiocondaBot please fetch artifacts` will post links to packages and docker containers built by the CI system. You can use this to test packages locally before merging.
>
> For members of the Bioconda project, the following command is also available:
>
>  * `@BiocondaBot please merge` will cause packages/containers to be uploaded and a PR merged. Someone must approve a PR first! This has the benefit of not wasting CI build time required by manually merging PRs.
>
> If you have questions, please post them in gitter or ping `@bioconda/core` in a comment (if you are not able to directly ping `@bioconda/core` then the bot will repost your comment and enable pinging).

### CODE REVIEW

File: recipes/iqtree/meta.yaml [hunk 1]
```diff
@@ -16,14 +16,15 @@ requirements:
   build:
     - {{ compiler('cxx') }}
     - cmake
+    - clangdev  # [osx]
```
Comment (thermokarst) [inline recipes/iqtree/meta.yaml]: Hey @andrewsanchez - maybe `host` still needs `clangdev`? Not sure why, but you could try putting this line back into the `host` section, below.

The [CI run](https://circleci.com/gh/bioconda/bioconda-recipes/98781?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-checks-link) had this little tidbit of a failure: `ld: library not found for -lcrt1.10.6.o`, which is why I am wondering if we need `clangdev` in `host`.

Comment (andrewsanchez) [reply recipes/iqtree/meta.yaml]: Good idea.  and right below that it says  `clang-4.0: error: linker command failed with exit code 1` which seems like it could be related to the last bit of info right before it exits:  `Linking CXX executable iqtree`

File: recipes/iqtree/build.sh [hunk 2]
```diff
@@ -1,11 +1,18 @@
...
-export CFLAGS="-I$PREFIX/include"
-export LDFLAGS="-L$PREFIX/lib"
+export CFLAGS="$CFLAGS -I$PREFIX/include"
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib"
 export CPATH=${PREFIX}/include
 
+if [ "$(uname)" == Darwin ]; then
+    export MACOSX_DEPLOYMENT_TARGET=10.7
+    export CMAKE_C_COMPILER="clang"
+    export CMAKE_CXX_COMPILER="clang++"
```
Comment (thermokarst) [inline recipes/iqtree/build.sh]: Can you toss this env var in here, too?

```bash
export CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY
```

This changes the way cmake performs its pre-run checks.

File: recipes/iqtree/meta.yaml [hunk 3]
```diff
@@ -6,11 +6,14 @@ package:
   version: {{ version }}
 
 build:
-  number: 0
+  number: 1
+  skip: true # [linux]
```
Comment (thermokarst) [inline recipes/iqtree/meta.yaml]: TODO: revert this platform pin (currently on to prevent spending time testing on linux, since this PR is attempting to address build issues specific to darwin).

File: recipes/iqtree/build.sh [hunk 4]
```diff
@@ -1,11 +1,19 @@
 #!/bin/bash
-export CFLAGS="-I$PREFIX/include"
-export LDFLAGS="-L$PREFIX/lib"
+export CFLAGS="$CFLAGS -I$PREFIX/include"
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib"
 export CPATH=${PREFIX}/include
 
+if [ "$(uname)" == Darwin ]; then
+    export MACOSX_DEPLOYMENT_TARGET=10.7
+    export CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY
```
Comment (thermokarst) [inline recipes/iqtree/build.sh]: Hmm, the latest circleci run implies that this change didn't get picked up - it appears to still be building a test executable...

File: recipes/iqtree/build.sh [hunk 5]
```diff
@@ -1,11 +1,19 @@
 #!/bin/bash
-export CFLAGS="-I$PREFIX/include"
-export LDFLAGS="-L$PREFIX/lib"
+export CFLAGS="$CFLAGS -I$PREFIX/include"
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib"
 export CPATH=${PREFIX}/include
 
+if [ "$(uname)" == Darwin ]; then
+    export MACOSX_DEPLOYMENT_TARGET=10.7
```
Comment (thermokarst) [inline recipes/iqtree/build.sh]: Please update this to 10.9 so that it matches the patchfile.

File: recipes/iqtree/build.sh (hunk 4 above)
Comment (thermokarst) [reply recipes/iqtree/build.sh]: You know, it might be the case that this flag can only be set with the `-D CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY` syntax, rather than as an env var.

Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: Let's see what happens in the next build.  The `patch` directive was not indented properly underneath `source` 🤦‍♂

File: bootstrap.py [hunk 6]
```diff
@@ -52,7 +52,7 @@
 
 # TODO: this mimics the override in the "common" job in .circleci/config.yaml
 with open('.circleci/common.sh', 'w') as fout:
-    fout.write("MINICONDA_VER=4.5.4\nBIOCONDA_UTILS_TAG=master\n")
+    fout.write("MINICONDA_VER=4.6.14\nBIOCONDA_UTILS_TAG=v0.16.11\n")
```
Comment (thermokarst) [inline bootstrap.py]: You might want to revert this, since it isn't part of the recipe.

Comment (andrewsanchez) [reply bootstrap.py]: Oops!  Thank you.

File: recipes/iqtree/build.sh (hunk 5 above)
Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: It was originally set to 10.9.  You asked to try 10.7 after that build failed.  Do you remember why you wanted it set to 10.7 on Friday?

File: recipes/iqtree/build.sh (hunk 4 above)
Comment (thermokarst) [reply recipes/iqtree/build.sh]: Looks like something is still off with the patch, can you double-check that?

File: recipes/iqtree/build.sh (hunk 5 above)
Comment (thermokarst) [reply recipes/iqtree/build.sh]: >  Do you remember why you wanted it set to 10.7 on Friday?

Yep! We temporarily tested against 10.7 because we noticed that half the build was using 10.7 while the other was using 10.9 - you changed this env var to 10.7 to see if we could get everything building against the same version of macos. That succeeded in getting everything building against 10.7, but didn't do anything to address the underlying problem. So then the next step was to try to build everything against 10.9. We started that by creating a patch to the CMake config to force that end to use 10.9, next we need to fix up this env var to use 10.9, too.

File: recipes/iqtree/macos.patch [hunk 7]
```diff
@@ -0,0 +1,21 @@
+diff --git a/CMakeLists.txt b/CMakeLists.txt
```
Comment (thermokarst) [inline recipes/iqtree/macos.patch]: Don't forget to drop this patchfile.

File: recipes/iqtree/build.sh [hunk 8]
```diff
@@ -1,11 +1,19 @@
 #!/bin/bash
-export CFLAGS="-I$PREFIX/include"
-export LDFLAGS="-L$PREFIX/lib"
+export CFLAGS="$CFLAGS -I$PREFIX/include"
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib"
 export CPATH=${PREFIX}/include
 
+if [ "$(uname)" == Darwin ]; then
+    export MACOSX_DEPLOYMENT_TARGET=10.9
```
Comment (thermokarst) [inline recipes/iqtree/build.sh]: Okay, as a quick sanity check, can you temporarily comment out this line, and the line in the recipe that applies the patch? I just want to make sure that forcing a 10.9 target is something necessary for this changeset.

File: recipes/iqtree/build.sh (hunk 4 above)
Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: Do you still want to run any tests re: `CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY` ?

File: recipes/iqtree/build.sh [hunk 9]
```diff
@@ -1,11 +1,19 @@
 #!/bin/bash
-export CFLAGS="-I$PREFIX/include"
-export LDFLAGS="-L$PREFIX/lib"
+export CFLAGS="$CFLAGS -I$PREFIX/include"
+export LDFLAGS="$LDFLAGS -L$PREFIX/lib"
 export CPATH=${PREFIX}/include
 
+if [ "$(uname)" == Darwin ]; then
+    # export MACOSX_DEPLOYMENT_TARGET=10.9
```
Comment (thermokarst) [inline recipes/iqtree/build.sh]: Thanks for commenting out, but it looks like you missed the patch, [below](https://github.com/bioconda/bioconda-recipes/pull/20859/files#diff-ad768d0ca78f5f0c5a55d9858164f237R16).

File: recipes/iqtree/build.sh (hunk 4 above)
Comment (thermokarst) [reply recipes/iqtree/build.sh]: Not just yet, let's focus on resolving one thing at a time right now - let's finish our 10.9 checks.

File: recipes/iqtree/build.sh (hunk 9 above)
Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: Ok, I will skip applying the patch as well.  Sorry, I missed that in your previous comment.

Comment (thermokarst) [reply recipes/iqtree/build.sh]: No worries, thanks!

File: recipes/iqtree/build.sh (hunk 8 above)
Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: Builds failed. Adding the patch and deployment target back in

File: recipes/iqtree/build.sh (hunk 4 above)
Comment (thermokarst) [reply recipes/iqtree/build.sh]: Okay, now that 10.9 is sorted out, can you check and see if this flag is necessary, too?

Comment (thermokarst) [reply recipes/iqtree/build.sh]: PS - this "parametric" development cycle is painful - sorry about that!

Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: To clarify, you just want to test with `export  CMAKE_TRY_COMPILE_TARGET_TYPE=STATIC_LIBRARY`? commented out?

Comment (andrewsanchez) [reply recipes/iqtree/build.sh]: All checks have passed without this flag.

**Comment by thermokarst** [comment]
@BiocondaBot please fetch artifacts

**Comment by andrewsanchez** [comment]
@BiocondaBot please add label

**Comment by davebx** [comment]
Thanks for all the work, @thermokarst, @andrewsanchez 

@BiocondaBot please merge

**Comment by davebx** [comment]
@BiocondaBot please merge
